### PR TITLE
Economy Expansion - "Black Market" Diner Merchants

### PR DESCRIPTION
**Pull request by DEAD-VICKY**
## About the PR
I got some requests to finish this. This is the completed and tested version. Previous version was a draft that was flagged ready by someone else.

.zip with all graphic & map files avaliable at https://easyupload.io/yxqb9c
Code might not compile without them.

Note: to dummy out trades, comment out their entries in traders.dm NOT commodities.dm

## Why's this needed?
First in a series of a planned updates to the economy system. Primary goal is to increase player interest in space cash. and by extention every system that uses cash. Mainly looking at various areas that haven't recieved content in a while and giving them something new.

Part 1: Static Traders (Black Market)
Part 2: QM Traders (Space Wizards, Gangsters & Scrappers)
Part 3: Vending Machines
Part 4: Random Event Merchants

## Changelog
```changelog
(u)DEADVICKY
(*)Added four (4) new merchant NPCs
(+)Gave all new merchants animated overworld & talking sprites
(+)Placed them in various spots around the Space Diner
(+)Placed them in various spots around Cogmap 2
(+)Placed them in various spots around Horizon
(+)Placed them in various spots around OSHAN
(*)One hundred and eleven (111) new commodities
```

### CODE REVIEW

**Comment by DEAD-VICKY** [comment]
> You”ve appeared to accidentally delete the ai frame kit commodities (likely when resolving merge conflicts). Also way you have blobs purchases set up means they wont work, at least in the way players expect. Currently it detects the pathing of the object, not what chems they contain. Theres going to have to be special code here that checks what reagents the containers contain, fortunately enough, requisitions already has this, id reccomend looking at how that does it and applying it to here.
Blob sells reagents and doesn't buy any because of this. I'd like to have him buy blood or some secret chems, but having him pay per unit is going to be complicated. Maybe in the future.
> Personally, i feel like having the traders buy several admin only (or at least for the chances of qms getting there hands on it and being able to sell it and meaningfully use it), is misleading to the qm that these things could be obtained and id prefer to see only obtainable things there.
There's legitimate cases where you want to sell admin-spawned gear. Because of the way the code works, including the option means it's either do this or outright lie to the player with a misleading description.

**Comment by ThatFiveGuys** [comment]
> Blob sells reagents and doesn't buy any because of this. I'd like to have him buy blood or some secret chems, but having him pay per unit is going to be complicated. Maybe in the future.
> There's legitimate cases where you want to sell admin-spawned gear. Because of the way the code works, including the option means it's either do this or outright lie to the player with a misleading description.

Apologies, misread blob code, however if a player wants to get admin gear out of the round, they can just use several of the many ways to delete items that already exist. Hell personally i feel as if encouraging qms to remove fun once in a blue moon items from a round would make some rounds a lot more dull.

**Comment by alexeric123** [comment]
Just something i saw from a quick overview, the hairball key doesn't exist anymore, at least not without admins.

File: code/modules/economy/commodity.dm [hunk 1]
```diff
@@ -2900,3 +2880,1072 @@
...
+	upperfluc = 3000
+	lowerfluc = -3000
+/datum/commodity/blackmarket/wrestler/nassa
+	comname = "NASSA space suit"
+	comtype = /obj/item/clothing/under/gimmick/blackstronaut
+	desc_buy = "I'd LOVE to get my hands on a couple sets of that prototype jumpsuit. WRESTLING IN THE COLD VACCUM OF SPACE! I'll pay a BUNCH for any you find in good condition!"
+	price = 6200
+	upperfluc = 800
+	lowerfluc = -800
+/datum/commodity/blackmarket/wrestler/uplink_telecrystal
```
Comment (Eagletanker) [inline code/modules/economy/commodity.dm:3896]: probably should have some way to be named after the current traitor buck of the round.

Also kinda useless considering traitors likely arent going to want to sell their hard earned antag dollarodoos for something they can easily get from just stealing the budget

File: code/modules/economy/commodity.dm [hunk 2]
```diff
@@ -2900,3 +2880,1072 @@
...
+	lowerfluc = -30000
+/datum/commodity/blackmarket/wrestler/shoecrafts
+	comname = "Custom Shoes"
+	comtype = /obj/item/clothing/shoes/crafted
+	desc_buy = "And any shoes YOU HAND CRAFT, I WILL BUY!"
+	price = 800
+	upperfluc = 600
+	lowerfluc = -600
+/datum/commodity/blackmarket/wrestler/katana_nukeop
+	comname = "Syndicate Commander's Sabre"
```
Comment (Eagletanker) [inline code/modules/economy/commodity.dm:3946]: Nuke ops rounds dont last this long

File: code/modules/economy/trader.dm [hunk 3]
```diff
@@ -1410,31 +1406,370 @@
...
+//////////////////////////////////////////////////////////////
+///███████ ███████ ██      ██      ██ ███    ██  ██████     //
+///██      ██      ██      ██      ██ ████   ██ ██       ██ //
+///███████ █████   ██      ██      ██ ██ ██  ██ ██   ███    //
+///     ██ ██      ██      ██      ██ ██  ██ ██ ██    ██ ██ //
+///███████ ███████ ███████ ███████ ██ ██   ████  ██████     //
+//////////////////////////////////////////////////////////////
+// Sells a full set of security gear, with random parts mixed in depending on what day it is.
+// Note: No headset, no taser
+		src.goods_sell += new /datum/commodity/blackmarket/surplus/baton(src)
```
Comment (Eagletanker) [inline code/modules/economy/trader.dm:1503]: probably shouldnt sell batons considering how powerful they are in combat.

File: code/modules/economy/trader.dm [hunk 4]
```diff
@@ -1410,31 +1406,370 @@
...
+///██   ██ ██    ██  ██  ██  ██ ████   ██ ██       ██ ///
+///██████  ██    ██   ████   ██ ██ ██  ██ ██   ███    ///
+///██   ██ ██    ██    ██    ██ ██  ██ ██ ██    ██ ██ ///
+///██████   ██████     ██    ██ ██   ████  ██████     ///
+/////////////////////////////////////////////////////////
+		src.goods_buy += new /datum/commodity/blackmarket/surplus/ak47(src)
+		src.goods_buy += new /datum/commodity/blackmarket/surplus/revolver(src)
+		src.goods_buy += new /datum/commodity/blackmarket/surplus/lasergun(src)
+		src.goods_buy += new /datum/commodity/blackmarket/surplus/phaser(src)
+//		src.goods_buy += new /datum/commodity/blackmarket/surplus/lawbringercell(src)
```
Comment (Eagletanker) [inline code/modules/economy/trader.dm:1526]: This cant be gotten. Lawbringer doesnt use a ejectable cell.

File: code/modules/economy/trader.dm [hunk 5]
```diff
@@ -1410,31 +1406,370 @@
...
+		pickupdialoguefailure = "Nothing in your cart, pal."
+//////////////////////////////////////////////////////////////
+///███████ ███████ ██      ██      ██ ███    ██  ██████     //
+///██      ██      ██      ██      ██ ████   ██ ██       ██ //
+///███████ █████   ██      ██      ██ ██ ██  ██ ██   ███    //
+///     ██ ██      ██      ██      ██ ██  ██ ██ ██    ██ ██ //
+///███████ ███████ ███████ ███████ ██ ██   ████  ██████     //
+//////////////////////////////////////////////////////////////
+		src.goods_sell += new /datum/commodity/blackmarket/shady/cursedmask(src)
+		src.goods_sell += new /datum/commodity/blackmarket/shady/omnitool(src)
```
Comment (Eagletanker) [inline code/modules/economy/trader.dm:1670]: Probably should check to see if someone is a syndie traitor, for all the gear that's syndicate in origin. Randomly having to adapt to some nerd baton stunning you down and then cluwneing you is not gonna be fun.

File: code/modules/economy/commodity.dm [hunk 6]
```diff
@@ -2900,3 +2880,1072 @@
...
+/datum/commodity/blackmarket/wrestler/chompskey
+	comname = "Chomp Key"
+	comtype = /obj/item/device/key/chompskey
+	desc_buy = "I need one of these URGENTLY!"
+	price = 20000
+	baseprice = 20000
+	upperfluc = 2000
+	lowerfluc = -2000
+/datum/commodity/blackmarket/wrestler/hairballkey
+	comname = "Hairball Key"
```
Comment (Eagletanker) [inline code/modules/economy/commodity.dm:3851]: as someone pointed out before this cannot be gotten

File: code/modules/economy/commodity.dm [hunk 7]
```diff
@@ -2900,3 +2880,1072 @@
...
+	upperfluc = 10
+	lowerfluc = -10
+/datum/commodity/blackmarket/wrestler/monkeyphoto
+	comname = "Official Ugly Monkey Photograph"
+	comtype = /obj/item/photo/uglymonkey
+	desc = "Highly collectible picture of a monkey! Each one comes with a number and a name!"
+	price = 500000
+	upperfluc = 1000000
+	lowerfluc = -1000000
+/datum/commodity/blackmarket/wrestler/zippo
```
Comment (Eagletanker) [inline code/modules/economy/commodity.dm:3676]: this appears twice

**Comment by wolfolotl** [comment]
having a trader that rewards and encourages greytiding seems like several self-antagging instances waiting to happen

**Comment by nefarious6th** [comment]
Mixed bag here - there's some ideas I do like in here (I think being able to outright buy space fungus, for example, would be neat).
I have reservations about some of the other stuff listed as goods to buy; namely the spare stun baton, breaching hammer, SecHUDs, NTSO gas mask, AK, and Predator.
Predator revolver is just a matter of high-powered traitor gear remaining traitor-exclusive.
NTSO mask is about identifiablity.
AK, Breaching Hammer, and Batom are all balance concerns.
SecHUDs are a special case I think since they compete with the ocular prostheses kit for traitor gear and the SecHUD cybereyes available from other vendors consistently, so I think here they may be getting outclassed and lose some uniqueness to them.
Others commented on some mechanical fixes already.
Only other comment is to reduce the ASCII art headers; I think the books.dm and cake.dm have good examples of space-efficient headers, if you need guidance.

File: code/modules/economy/commodity.dm [hunk 8]
```diff
@@ -2900,3 +2880,1072 @@
 	price = 21000
 	upperfluc = 20000
 	lowerfluc = -28500
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+//  ______   _        _______  _______  _                _______  _______  _______  _        _______ _________ //
+// (  ___ \ ( \      (  ___  )(  ____ \| \    /\        (       )(  ___  )(  ____ )| \    /\(  ____ \\__   __/ //
```
Comment (Eagletanker) [inline code/modules/economy/commodity.dm:2886]: can you please drop the ASCII art? It makes it harder to read whats going on

**Review by Eagletanker — COMMENTED** [review]
Just some small thoughts as a non dev on this.

File: code/modules/economy/commodity.dm (hunk 8 above)
Comment (DEAD-VICKY) [reply code/modules/economy/commodity.dm:2886]: no

**Comment by NightmareChamillian** [comment]
This is showing MASSIVE improvements over the dumpsterfire that was the first iteraction of this. Some suggestions nonethless:

For the surplus boots/helmet/mask, I strongly, strongly suggest that they get randomized, for the sake of making it more exciting and also MUCH harder to reliably impersonate sec.
The stun baton should be the securitron one, but slightly cheaper. I don't think being able to just buy a rechargable stun baton is worth it, regardless of the limitations, this also forces players to choose whether or not they want to carry multiple.
The vest for sale should also have some RNG to give a chance of being shitty, and be more expensive. A lot of what the surplus guy is selling just seems to be a secondhand security crate that you don't have to unlock, which is bad. 

I think the blob should sell more stage 1-2 chems, like acetone, ephedrine, and acetic acid. Things that are incharacter, but also a pain in the ass to get.

The syndicate stuff should be WAY more expensive or have some other failure point. This is essentially just free traitor stuff. However, I do think it should sell expensive 9mm bullets in batches of 4-6, which would be something that's not basically free traitor gear with how quickly people get money, but still useful and desirable

Again, I think the day dependent things-in this case the t shirts-should be randomized. It adds excitement.

**Comment by DEAD-VICKY** [comment]
not doing randomisation

**Comment by Azrun** [comment]
This does not appear to be ready to review.  Continuous integration is 😢

**Comment by TobleroneSwordfish** [comment]
> Code might not compile without them.

You might want to include them then, or the CI is going to be sad and it's hard to know whether any of this code works.

**Comment by DEAD-VICKY** [comment]
> > Code might not compile without them.
> 
> You might want to include them then, or the CI is going to be sad and it's hard to know whether any of this code works.

they're included in the link. cant upload directly to github because my account got mass reported as a spammer a few weeks ago so you'll have to test it yourself

**Comment by DEAD-VICKY** [comment]
> I think the blob should sell more stage 1-2 chems, like acetone, ephedrine, and acetic acid. Things that are incharacter, but also a pain in the ass to get.

Agreed, leaving more chems for a future patch though. Chems and especially low-tier ones in many recipes can combo in unpredictible ways so each one needs a bit of testing, a lot of unassuming chemicals become insane death/healing mixes when combined with vending machine items. Liquid oxygen and sulphuric acid were two I tested that were very powerful. Also I was having trouble with blob merchant dialogue and I might rewrite them from scratch. If any commoditity doesn't have a unique line of dialogue it leaves an ugly error message so having enough decent lines takes effort.

**Comment by Vaskritaya** [comment]
I do not speak for the developers, but I have my doubts that including map files and graphics in a separate download is acceptable. On top of the added workload for any dev trying to add this, it makes various mechanical advantages of using github, such as diff checks,  runtime checks, and ensuring that the pr only touches things related to the pr totally useless. 

Out of curiosity, would you know why you are able to push changes to things such as commodity.dm, but unable to push changes to things such as oshan.dmm? Any error messages you might be seeing? Could potentially be a problem with vscode or similar.

File: code/modules/economy/commodity.dm [hunk 9]
```diff
@@ -2900,3 +2880,1072 @@
...
+		reagents.add_reagent("booster_enzyme", 10)
+/obj/item/reagent_containers/glass/flask/oil
+	name = "Flask of Oil"
+	desc = "A flask with a label on it. The label says 'OIL'."
+	icon_state = "bottle3"
+	amount_per_transfer_from_this = 10
+	initial_volume = 50
+	New()
+		..()
+		reagents.add_reagent("triplepiss", 50)
```
Comment (Vaskritaya) [inline code/modules/economy/commodity.dm:3256]: The oil flask seems to create triplepiss. I'm assuming you forgot to change it to oil when copypasting the triplepiss flask.

**Comment by mordent-goonstation** [comment]
Fails to compile, we're not going to manually go and add in some items from a .zip file. I'm unsure why you are unable to add in changes to those files but you can to the .dm files.

My own opinion is that the ASCII acting as file sectioning is a a hard "no" and is a strong indicator that if you think it is required that the file should maybe be split into multiple files.

This PR is unmergeable. I have not looked at the actual contents/changes.

**Comment by DEAD-VICKY** [comment]
> I do not speak for the developers, but I have my doubts that including map files and graphics in a separate download is acceptable. On top of the added workload for any dev trying to add this, it makes various mechanical advantages of using github, such as diff checks, runtime checks, and ensuring that the pr only touches things related to the pr totally useless.
> 
> Out of curiosity, would you know why you are able to push changes to things such as commodity.dm, but unable to push changes to things such as oshan.dmm? Any error messages you might be seeing? Could potentially be a problem with vscode or similar.

Simply don't have the permissions to update non-text files. I need a member of the coding team to copy paste them from my zip.

> Fails to compile, we're not going to manually go and add in some items from a .zip file. I'm unsure why you are unable to add in changes to those files but you can to the .dm files.
> 
> My own opinion is that the ASCII acting as file sectioning is a a hard "no" and is a strong indicator that if you think it is required that the file should maybe be split into multiple files.
> 
> This PR is unmergeable. I have not looked at the actual contents/changes.

The files in the zip are sorted by folder and copying them to github will take less than five minutes.

File: code/modules/economy/trader.dm (hunk 5 above)
Comment (DEAD-VICKY) [reply code/modules/economy/trader.dm:1670]: it's an acid mask not a cluwne mask

**Comment by ThatFiveGuys** [comment]
Unsure how youll trying to add the pictures but youll want to add them as a dmi file, rough instructions can be found here: https://hackmd.io/@goonstation/docs/%2F%40goonstation%2Fdev.

**Pull request closed without merging**